aten/src/ATen/VmapTransforms.cpp

```diff
@@ -94,33 +94,128 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
+}
+
+// Given a Tensor or a BatchedTensor, returns the underlying physical tensor
+// with all vmapped dimensions permuted to the front, if they exist, and a
+// bitset of vmap levels that were present in the tensor.
+static std::pair<Tensor,std::bitset<kVmapNumLevels>>
+getPhysicalTensorAndLevels(const Tensor& self) {
+  auto* batched = maybeGetBatched(self);
+  if (batched) {
+    return {permuteBatchDimsToFront(batched), createLevelsBitset(batched->bdims())};
   }
-  if (reference_sizes.size() < num_batch_dims) {
-    return false;
+  return {self, 0};
+}
+
+// Given a Tensor or a BatchedTensor, creates a physical view of the tensor
+// such that it has a batch dimension for each level in `requested_levels`
+// and `requested_example_dim` number of non-batch-dimensions.
+//
+// This function is useful in preparing physical views on tensors that can
+// then be passed into broadcasting operations. For example, when adding
+// two BatchedTensors of sizes [B0, 3] and [B0, B1, 2, 3], where the Bi are the
+// batch dimensions, we must align the batch dimensions and non-batch-dimensions
+// (henceforth referred to as the "example" dimensions) separately to produce
+// tensors of size [B0, 1, 1, 3] and [B0, B1, 2, 3] so that they can be added.
+//
+// Here's a direct example of using alignBatchDimsAtFront on the above two tensors.
+//
+// 1) alignBatchDimsAtFront([B0, 3], requested_levels={0, 1}, requested_example_dim=2)
+// returns a physical view of size [B0, 1, 1, 3] by adding an extra dimension for
+// level 1 and another extra dimension to pad the example dimensions to 2.
+//
+// 2) alignBatchDimsAtFront([B0, B1, 2, 3], requested_levels={0, 1}, requested_example_dim=2)
+// returns a physical view of size [B0, B1, 2, 3]
+static Tensor alignBatchDimsAtFront(
+    const Tensor& self,
+    std::bitset<kVmapNumLevels> requested_levels,
+    int64_t requested_example_dim) {
+  Tensor physical_tensor;
+  std::bitset<kVmapNumLevels> tensor_levels;
+  std::tie(physical_tensor, tensor_levels) = getPhysicalTensorAndLevels(self);
+
+  TORCH_INTERNAL_ASSERT(
+    (tensor_levels | requested_levels) == requested_levels,
+    "`requested_levels` must be a superset of `self`'s levels");
+
+  auto physical_sizes = physical_tensor.sizes();
+
+  auto tensor_example_dim = physical_sizes.size() - /*num_batch_dims*/tensor_levels.count();
+  TORCH_INTERNAL_ASSERT(tensor_example_dim <= requested_example_dim);
+
+  std::vector<int64_t> aligned_sizes(requested_levels.count() + requested_example_dim, 1);
+
+  // align the example dims (non-bdims dims) first
+  // aligned_sizes[-tensor_example_dim:] = tensor_sizes[-tensor_example_dim:]
+  std::copy(
+      physical_sizes.rbegin(),
+      physical_sizes.rbegin() + tensor_example_dim,
+      aligned_sizes.rbegin());
+
+  // align the bdims
+  int64_t level = 0;
+  int64_t dim = 0;
+  int64_t tensor_dim = 0;
+  for (int64_t level = 0; level < kVmapNumLevels; level++) {
```

Comment:
Something that made this a little hard to understand for me was the fact that you were iterating over *all* 64 levels.  Intuitively to me, I would have expected to process only the union of levels which actually existed (aka, the loop counter ought to be `dim`). This is not a suggestion to change the implementation per se; this may very well be the most convenient way to go about doing it as things are written, but I did have to do a little mental leap here to figure out what was going on.

Reply:
Oh, that's a really good point. When I originally wrote this code I thought it was more convenient to do it this way. But now that I think about it more, it's easier to write this by iterating over `dim`.